transports/noise/src/io/framed.rs

```diff
@@ -106,270 +133,70 @@ impl<T> NoiseFramed<T, snow::HandshakeState> {
-                    }
-                    *off += n;
-                    if len == *off {
-                        trace!("read: decrypting {} bytes", len);
-                        this.decrypt_buffer.resize(len, 0);
-                        if let Ok(n) = this
-                            .session
-                            .read_message(&this.read_buffer, &mut this.decrypt_buffer)
-                        {
-                            this.decrypt_buffer.truncate(n);
-                            trace!("read: payload len = {} bytes", n);
-                            this.read_state = ReadState::Ready;
-                            // Return an immutable view into the current buffer.
-                            // If the view is dropped before the next frame is
-                            // read, the `BytesMut` will reuse the same buffer
-                            // for the next frame.
-                            let view = this.decrypt_buffer.split().freeze();
-                            return Poll::Ready(Some(Ok(view)));
-                        } else {
-                            debug!("read: decryption error");
-                            this.read_state = ReadState::DecErr;
-                            return Poll::Ready(Some(Err(io::ErrorKind::InvalidData.into())));
-                        }
-                    }
-                }
-                ReadState::Eof(Ok(())) => {
-                    trace!("read: eof");
-                    return Poll::Ready(None);
-                }
-                ReadState::Eof(Err(())) => {
-                    trace!("read: eof (unexpected)");
-                    return Poll::Ready(Some(Err(io::ErrorKind::UnexpectedEof.into())));
-                }
-                ReadState::DecErr => {
-                    return Poll::Ready(Some(Err(io::ErrorKind::InvalidData.into())))
-                }
+impl asynchronous_codec::Decoder for Codec<snow::HandshakeState> {
+    type Error = io::Error;
+    type Item = proto::NoiseHandshakePayload;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        match self.decode_bytes(src)? {
+            Some(bytes) => {
+                let mut reader = BytesReader::from_bytes(&bytes[..]);
+                let pb = proto::NoiseHandshakePayload::from_reader(&mut reader, &bytes[..])
+                    .map_err(|_| {
+                        io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            "Failed decoding handshake payload",
+                        )
+                    })?;
+                Ok(Some(pb))
             }
+            None => Ok(None),
         }
     }
 }
 
-impl<T, S> futures::sink::Sink<&Vec<u8>> for NoiseFramed<T, S>
-where
-    T: AsyncWrite + Unpin,
-    S: SessionState + Unpin,
-{
+impl asynchronous_codec::Encoder for Codec<snow::TransportState> {
     type Error = io::Error;
+    type Item<'a> = &'a Vec<u8>;
```

Comment:
I think a slice would be more idiomatic:

```suggestion
    type Item<'a> = &'a [u8];
```